### PR TITLE
feat(build): add github action

### PR DESCRIPTION
**Pull request by amitbhatt818**
This commit adds github actions file which performs the following:
1. Install eksctl
2. Configure AWS credentials from secrets
3. Create EKS cluster using eksctl
4. Checkout this repo
5. Install DOP on cluster
6. Cluster cleanup

## Positive scenario(s) with output
[log.txt](https://github.com/mayadata-io/kubera-api-testing/files/5153850/log.txt)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

Signed-off-by: Amit Bhatt <amit.bhatt@mayadata.io>

### CODE REVIEW

File: .github/workflows/action.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    - name: Configure AWS credentials from secrets
+      uses: aws-actions/configure-aws-credentials@v1
+      with:
+        aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+        aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+        aws-region: ${{ secrets.AWS_DEFAULT_REGION }}
+        
+    - name: Create EKS cluster using eksctl
+      run: |
+        eksctl create cluster --name=test --node-type=t2.xlarge
```
Comment (rohan2794) [inline .github/workflows/action.yml]: can we update cluster name to something like `dop_cluster` or `kubera_cluster` or any other which you think will be more suitable

File: .github/workflows/action.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    
+    - name: Install DOP on cluster
+      run: |
+        chmod +x ./dop.sh
+        ./dop.sh
+
+    - name: Cluster cleanup
+      run: |
+        eksctl delete cluster --name=test
+
```
Comment (rohan2794) [inline .github/workflows/action.yml]: can we also verify whether cluster is deleted or not after `eksctl delete cluster --name=test` run

File: dop.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+## Install helm
```
Comment (AmitKumarDas) [inline dop.sh]: Let us have EKS & DOP & run test cases as a single script
Lets rename this to `verify_dop_on_eks.sh`
GitHub Actions will just trigger this script.

File: dop.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3
+chmod 700 get_helm.sh
+./get_helm.sh
+helm version
+
+## Creates secret
```
Comment (AmitKumarDas) [inline dop.sh]: Please explain what this secret is all about.

File: dop.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+kubectl create secret docker-registry directoronprem-registry-secret --docker-server=registry.mayadata.io --docker-username=${{ DOCKER_USERNAME }} --docker-password=${{ DOCKER_PASSWORD }}
```
Comment (AmitKumarDas) [inline dop.sh]: Indent this line of code & limit the line length to max 80 chars. We should not do horizontal scolling.

File: dop.sh (hunk 3 above)
Comment (AmitKumarDas) [inline dop.sh]: Please put appropriate shebang & set options.

Comment (AmitKumarDas) [inline dop.sh]: Please add linting check to verify the correctness of this script.
Refer shellcheck!
- https://github.com/koalaman/shellcheck

File: dop.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+kubectl get secret
+
+## Install Dop
```
Comment (AmitKumarDas) [inline dop.sh]: Can this be explained in comments?
Why is this required?

**Review by prabhatkumarthakur — COMMENTED** [review]
you need to remove issue template and PR template from .github directory

File: .github/workflows/action.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+#       uses: actions-contrib/golangci-lint@v1
+
+    - name: Configure AWS credentials from secrets
+      uses: aws-actions/configure-aws-credentials@v1
+      with:
+        aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+        aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+        aws-region: ${{ secrets.AWS_DEFAULT_REGION }}
+    
+    - name: Create EKS cluster using eksctl and Install kubera on cluster
```
Comment (AmitKumarDas) [inline .github/workflows/action.yml]: Can we break this into multiple actions i.e. scripts
one to install EKS
other one to install DOP on Kubernetes & then run preflight checks

File: .github/workflows/action.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      run: |
+        chmod +x ./utils/verify_kubera_on_eks.sh
+        ./utils/verify_kubera_on_eks.sh
+      env:
+        DOCKER_USER: ${{ secrets.DOCKER_USERNAME }}
+        DOCKER_PASS: ${{ secrets.DOCKER_PASSWORD }}
+
+    - name: Cluster and kubera cleanup
+      run: |
+        chmod +x ./utils/cleanup.sh
```
Comment (AmitKumarDas) [inline .github/workflows/action.yml]: can we break this into uninstall DOP & run postflight checks
uninstall EKS

File: .github/workflows/action.yml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    - name: Run Shellcheck
+      uses: bewuethr/shellcheck-action@v2
+
+##     TODO:: There is no go-file in repo. Thatswhy we are
+##     commenting this lines for now. 
+#
+#     - name: Run golangci-lint
+#       uses: actions-contrib/golangci-lint@v1
+
+    - name: Configure AWS credentials from secrets
```
Comment (AmitKumarDas) [inline .github/workflows/action.yml]: What happens if there is any error after successful installation of EKS?
Does GitHub action delete the cluster?

File: utils/cleanup.sh [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Comment (AmitKumarDas) [inline utils/cleanup.sh]: Let us rename this folder to eks

File: .github/workflows/action.yml (hunk 7 above)
Comment (amitbhatt818) [reply .github/workflows/action.yml]: Yes we can do that but install EKS has only 3 commands @AmitKumarDas . Is it fine to create separate file for that

File: .github/workflows/action.yml (hunk 8 above)
Comment (amitbhatt818) [reply .github/workflows/action.yml]: We have only one command in postflight check @AmitKumarDas

File: .github/workflows/action.yml (hunk 9 above)
Comment (amitbhatt818) [reply .github/workflows/action.yml]: No, it will not delete I will push one more commit where it will delete cluster if anything fails @AmitKumarDas

**Comment by amitbhatt818** [comment]
> you need to remove issue template and PR template from .github directory

Why we need to remove these files @prabhatkumarthakur any specific reason

**Review by prabhatkumarthakur — COMMENTED** [review]
I am asking to remove the issue template and PR template directory since it is already present in the master branch.

File: .github/workflows/action.yml (hunk 9 above)
Comment (amitbhatt818) [reply .github/workflows/action.yml]: I have pushed commit for this now if any failure happens then also clean step will get executed @AmitKumarDas

File: .github/workflows/action.yml [hunk 11]
```diff
@@ -0,0 +1,48 @@
+name: Github action to create EKS cluster using eksctl and install DOP on it
+on: [push]
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Install eksctl
+      run: |
+        curl --silent --location "https://github.com/weaveworks/eksctl/releases/latest/download/eksctl_$(uname -s)_amd64.tar.gz" | tar xz -C /tmp
```
Comment (AmitKumarDas) [inline .github/workflows/action.yml]: can we indent this command

File: .github/workflows/action.yml [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        sudo mv /tmp/eksctl /usr/local/bin
+        sleep 10
```
Comment (AmitKumarDas) [inline .github/workflows/action.yml]: can we avoid sleep

File: .github/workflows/action.yml (hunk 7 above)
Comment (AmitKumarDas) [reply .github/workflows/action.yml]: Yes. Lets separate them.

File: .github/workflows/action.yml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      uses: bewuethr/shellcheck-action@v2
+
+##     TODO:: There is no go-file in repo. Thatswhy we are
+##     commenting this lines for now. 
+#
+#     - name: Run golangci-lint
+#       uses: actions-contrib/golangci-lint@v1
+
+    - name: Configure AWS credentials from secrets
+      uses: aws-actions/configure-aws-credentials@v1
```
Comment (AmitKumarDas) [inline .github/workflows/action.yml]: What does this action do?
Can we write in comments

Comment (AmitKumarDas) [reply .github/workflows/action.yml]: I am hesitant to have so much of intelligence in github actions.
Can you install EKS without github actions?
can you install EKS from your laptop using our scripts?

File: eks/verify_kubera_on_eks.sh [hunk 14]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+ 
+set +e
+
+## Creating AWS cluster with two nodes and noty type t2.xlarge
+## in region us-east-2
+eksctl create cluster --name=kubera-cluster --node-type=t2.xlarge
+kubectl get nodes -owide
+sleep 200
```
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: Can we avoid all kinds of sleep?

File: eks/verify_kubera_on_eks.sh [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+kubectl get secret
+
+## To access kubera installation approved solution is to use an external Load Balancer 
+## with DNS to point to the hostname you are using for server.url
+IP=$(sudo kubectl get nodes -o wide --no-headers | awk '{print $6}' | head -n 1)
+URL=http://$IP
+helm repo add kubera https://charts.mayadata.io/
+helm repo update
+helm install kubera kubera/kubera-charts --set server.url="$URL"
+sleep 300
```
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: Can we avoid sleep

File: .github/workflows/action.yml (hunk 12 above)
Comment (amitbhatt818) [reply .github/workflows/action.yml]: removed

File: .github/workflows/action.yml (hunk 13 above)
Comment (amitbhatt818) [reply .github/workflows/action.yml]: created a seperate file. Now it is completely independent with github actions @AmitKumarDas

File: eks/verify_kubera_on_eks.sh (hunk 15 above)
Comment (amitbhatt818) [reply eks/verify_kubera_on_eks.sh]: No, we can't because we need sleep so that all the pods should come in `Running` state @AmitKumarDas . We can remove this sleep once we have sanity check for all the kubera components.

**Review by rohan2794 — APPROVED** [review]
/lgtm

File: eks/install_eks.sh [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+eksctl version
+
+## Configure AWS credentials 
+aws configure set aws_access_key_id "$ACCESS_KEY"
+aws configure set aws_secret_access_key "$SECRET_KEY"
+aws configure set default.region "$REGION"
+
+## Creating AWS cluster with two nodes and noty type t2.xlarge
+## in region us-east-2
+eksctl create cluster --name=kubera-cluster --node-type=t2.xlarge
```
Comment (AmitKumarDas) [inline eks/install_eks.sh]: Can we make use of declarative way i.e. yaml to install EKS of desired nodes, ram, cpu, etc?

File: eks/install_eks.sh [hunk 17]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+ 
+set +e
+
+## Install eksctl
```
Comment (AmitKumarDas) [inline eks/install_eks.sh]: can we add echo statements to represent logging

File: eks/verify_kubera_on_eks.sh [hunk 18]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: can we rename this to install_kubera_on_eks

File: eks/verify_kubera_on_eks.sh [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+ 
+set +e
+
+## Install helm
```
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: add echo as log

File: eks/verify_kubera_on_eks.sh [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+curl -fsSL -o get_helm.sh \
+https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3
+
+chmod 700 get_helm.sh
+./get_helm.sh
+helm version
+
+## Creates secret this secret will help us to pull images from registry
+kubectl create secret docker-registry directoronprem-registry-secret \
```
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: are these two secrets or a single secret.
add echo to log

File: eks/cleanup.sh [hunk 21]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Comment (AmitKumarDas) [inline eks/cleanup.sh:1]: can we refer to these scripts to write logs, indentations, script options like `set-eou pipefail`, etc.
https://github.com/raspbernetes/k8s-gitops/blob/master/scripts/flux.sh

File: eks/verify_kubera_on_eks.sh (hunk 19 above)
Comment (AmitKumarDas) [reply eks/verify_kubera_on_eks.sh]: move helm near to where it is being used

File: eks/verify_kubera_on_eks.sh [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+--docker-server=registry.mayadata.io \
+--docker-username="$DOCKER_USER" --docker-password="$DOCKER_PASS"
+kubectl get secret
+
+## To access kubera installation approved solution is to use an external Load Balancer 
+## with DNS to point to the hostname you are using for server.url
+IP=$(sudo kubectl get nodes -o wide --no-headers | awk '{print $6}' | head -n 1)
+URL=http://$IP
+helm repo add kubera https://charts.mayadata.io/
```
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: i find this style to be more readable
```sh
helm repo add fluxcd https://charts.fluxcd.io

helm template fluxcd/flux \
    --name-template=default \
    --namespace=flux \
    --version=1.5.0 \
    --values=config/flux/values.yaml > flux.yaml

[[ -f flux.yaml ]] && kubectl apply -f flux.yaml
```

File: eks/cleanup.sh [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+ 
+set +e
+
+## Dop cleanup
+helm uninstall kubera
```
Comment (AmitKumarDas) [inline eks/cleanup.sh]: I find following to be more readable. Whats your take?
```sh
if [[ -f "flux.yaml" && $CLEAN == true ]]; then
    echo -e "\nCleaning up manifests."
    echo "Set CLEAN=false if you wish for this not to occur."
    rm -rf flux.yaml
fi
```

File: eks/verify_kubera_on_eks.sh (hunk 22 above)
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: In order to user particular branches of this private repo, we might need to provide authentication.
e.g. https://github.com/mayadata-io/kubera-charts/tree/2.0
https://github.com/mayadata-io/kubera-charts/tree/master
Above will need us to provide auth

File: eks/verify_kubera_on_eks.sh [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+helm repo update
+helm install kubera kubera/kubera-charts --set server.url="$URL"
```
Comment (AmitKumarDas) [inline eks/verify_kubera_on_eks.sh]: Let us make use of helm linting
https://helm.sh/docs/helm/helm_lint/

File: eks/verify_kubera_on_eks.sh (hunk 22 above)
Comment (AmitKumarDas) [reply eks/verify_kubera_on_eks.sh]: Please check if we can avoid git clone of chart repo & just use helm to get the stuff done

File: eks/cleanup.sh (hunk 23 above)
Comment (rohan2794) [reply eks/cleanup.sh]: @AmitKumarDas what is `flux.yaml` here

File: eks/verify_kubera_on_eks.sh (hunk 20 above)
Comment (amitbhatt818) [reply eks/verify_kubera_on_eks.sh]: Single secret

File: eks/cleanup.sh (hunk 23 above)
Comment (AmitKumarDas) [reply eks/cleanup.sh]: @rohan2794 I was referring to code syntax & not flux

File: eks/cleanup.sh [hunk 25]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+ 
+CLEAN=${CLEAN:-true}
+
+if [[ $CLEAN == true ]]; then
+    echo -e "\nCleaning up Kubera."
+    echo "Set CLEAN=false if you wish for this not to occur."
+    helm uninstall kubera -n kubera
+    echo "Delete PVC"
```
Comment (AmitKumarDas) [inline eks/cleanup.sh]: The logs still show up as cluttered.
In addition this is "Delete all PVCs from kubera namespace"

File: eks/cleanup.sh [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+    kubectl delete pvc --all -n kubera
```
Comment (AmitKumarDas) [inline eks/cleanup.sh]: Logs give error on deleting PVCs. Are these actually deleted?

File: eks/cleanup.sh [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+    echo "Delete ns maya-system"
+    kubectl delete ns maya-system  
+    echo "Delete cluster"
+    eksctl delete cluster --name=kubera-cluster 
+    eksctl get cluster --region=us-east-2
```
Comment (AmitKumarDas) [inline eks/cleanup.sh:27]: Why are we mixing up the echo log with multiple actions?

File: eks/install_eks.sh [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+
+sudo mv /tmp/eksctl /usr/local/bin
+eksctl version
+
+echo -e "Configure AWS credentials"
+aws configure set aws_access_key_id "$ACCESS_KEY"
+aws configure set aws_secret_access_key "$SECRET_KEY"
+aws configure set default.region "$REGION"
+
+echo -e "Creating AWS cluster with two nodes and node type t2.xlarge in region us-east-2"
```
Comment (AmitKumarDas) [inline eks/install_eks.sh]: No need to mention the details in the logs.

File: eks/install_kubera_on_eks.sh [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+  echo -e "\nCloning $KUBERA_BRANCH branch of kubera-charts repo\n"
+  git clone -b "$KUBERA_BRANCH" https://"$GIT_USERNAME":"$GIT_PASSWORD"@github.com/mayadata-io/kubera-charts.git
+  cd kubera-charts
+  echo -e "Set release version,openebsRCEnable,maxMemberCountInOneProject and URL"
+  echo -e "\nInstalling Kubera using kubera-charts/values.yaml"
+  helm install --namespace kubera kubera \
+  --set server.url="$URL" \
+  --set server.release="$RELEASE" \
+  --set server.openebsRCEnable=true \
+  --set server.maxMemberCountInOneProject=50 -f values.yaml .
```
Comment (AmitKumarDas) [inline eks/install_kubera_on_eks.sh:45]: I guess values.yaml will taken up automatically

File: eks/install_kubera_on_eks.sh [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+  cd ..
+
+  echo -e "Changed values in values.yaml\n"
+  helm get values kubera -n kubera > value.yaml
```
Comment (AmitKumarDas) [inline eks/install_kubera_on_eks.sh]: Let us not print values, since its increasing the logs & hence clutter

File: eks/install_kubera_on_eks.sh [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+  helm install kubera kubera/kubera-charts \
+  --set server.url="$URL" \
+  --set server.release="$RELEASE" \
+  --set server.maxMemberCountInOneProject=100 \
+  --set server.openebsRCEnable=true \
+  -n kubera
+
+  echo -e "Changed values in values.yaml\n"
+  helm get values kubera -n kubera > value.yaml
+  cat value.yaml
```
Comment (AmitKumarDas) [inline eks/install_kubera_on_eks.sh]: same as previous comment

File: eks/cleanup.sh (hunk 26 above)
Comment (amitbhatt818) [reply eks/cleanup.sh]: There is one typo in command thatswhy it was giving error. I will fix this in next commit. @AmitKumarDas

**Review by AmitKumarDas — APPROVED** [review]
